src/enclave/Enclave/NonObliviousSortMergeJoin.cpp

```diff
@@ -128,20 +156,27 @@ void non_oblivious_sort_merge_join(
            
           // Reset primary_unmatched_rows
           primary_unmatched_rows.clear();
-          auto new_primary_unmatched_rows_buffer = new_primary_unmatched_rows.output_buffer();
-          RowReader new_primary_unmatched_rows_reader(new_primary_unmatched_rows_buffer.view());
-          while (new_primary_unmatched_rows_reader.has_next()) {
-            primary_unmatched_rows.append(new_primary_unmatched_rows_reader.next());
-          }
+          write_output_rows(new_primary_unmatched_rows, primary_unmatched_rows, join_type);
         }
       }
     }
   }
 
-  if (join_type == tuix::JoinType_LeftSemi) {
-    write_output_rows(primary_matched_rows, w);
-  } else if (join_type == tuix::JoinType_LeftAnti) {
-    write_output_rows(primary_unmatched_rows, w);
+  switch (join_type) {
+    case tuix::JoinType_LeftSemi:
+      write_output_rows(primary_matched_rows, w, join_type);
+      break;
+    case tuix::JoinType_LeftAnti:
+      write_output_rows(primary_unmatched_rows, w, join_type);
+      write_output_rows(previous_primary_unmatched_rows, w, join_type);
```

Comment:
If the input data is very big, then won't `previous_primary_unmatched_rows` potentially have to buffer a lot of data, and it will have to write out to `w` again? Can you write out to `w` directly earlier?

Reply:
The issue with doing it earlier is that `last_foreign_row.get()` can be null. This happens when a new primary group is encountered without any foreign row being encoounted first at all.

Reply:
For outer joins, can you put the dummy row first instead of last?

Reply:
that worked, thanks